include/ginkgo/core/config/config.hpp

```diff
@@ -0,0 +1,114 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_CONFIG_CONFIG_HPP_
+#define GKO_PUBLIC_CORE_CONFIG_CONFIG_HPP_
+
+
+#include <map>
+#include <string>
+#include <unordered_map>
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/config/type_descriptor.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+
+
+namespace gko {
+namespace config {
+
+class registry;
+
+
+class pnode;
+
+
+/**
+ * parse is the main entry point to create an Ginkgo object based on
+ * some file configuration. It reads a configuration stored as a property tree
+ * and creates the desired type.
+ *
+ * The configuration needs to specify the resulting type by the field:
+ * ```
+ * type: "some_supported_ginkgo_type"
+ * ```
+ * The result will be a deferred_factory_parameter, which can be thought of as
+ * an intermediate step before a LinOpFactory. Providing the result an Executor
+ * through the function `.on(exec)` will then create the factory with the
+ * parameters as defined in the configuration.
+ *
+ * Given a configuration that is defined as
+ * ```
+ * type: "solver::Gmres",
+ * krylov_dim: 20,
+ * stop: [
+ *   {iteration: 10},
+ *   {residual_norm: 1e-6}
+ * ]
+ * ```
+ * then passing it to this function like this:
+ * ```c++
+ * auto gmres_factory = build_from_config(config, context);
+ * ```
+ * will create a factory for a GMRES solver, with the parameters `krylov_dim`
+ * set to 20, and a combined stopping criteria, consisting of an Iteration
+ * criteria with maximal 10 iterations, and a ResidualNorm criteria with a
+ * reduction factor of 1e-6.
+ *
+ * By default, the factory will use the value type double, and index type
+ * int32 when creating templated types. This can be changed by passing in a
+ * type_descriptor. For example:
+ * ```c++
+ * auto gmres_factory = build_from_config(config, context,
+ *                                        make_type_descriptor<float, int>());
+ * ```
+ * will lead to a GMRES solver that uses `float` as its value type.
+ * Additionally, the config can be used to set these types through the fields:
+ * ```
+ * value_type: "some_value_type"
+ * index_type: "some_index_type"
```

Comment:
If we need to go for snake case, I will prefer the one with `_` because it is what we usually did in other codes.
I take a look at the other template types roughly. 
Except for ValueType and IndexType, `preconditioner::Ic` and `preconditioner::Ilu` uses some CamelName like `LSolverType` or something else.
I think that the CamelName can also help for distinguishing the template and parameter, but I also wonder whether there is a template type using snake_case.
I currently write the doucument and update the code for snake case.

Reply:
> I am struggling to imagine a case where it makes sense to specify a different index_type - do we need this as a configuration parameter?

this is still unadressed - since we need to statically know the `index_type` in any case, there is no purpose in configuring something that only has a single valid value. I think we should require the `index_type` to be set for `Multigrid` and solvers, and otherwise fail.